### ISSUE TITLE
"Running vertex group: influence" causes lag when called too often in a script

### ISSUE THREAD

**Issue by MileyHollenberg** [TheDuckCow/MCprep#92]
# Description
I'm currently trying to import 4560 chunks from a Minecraft world (exported via jMC2obj per chunk through a different script) into Blender and exporting them as an FBX after running `perp_materials` and `meshswap`. I am doing this per chunk as anything larger than 1 full chunk usually doesn't go well with `meshswap`. When running my Python script I notice that it gets slower each chunk when it reaches the following bit of code https://github.com/TheDuckCow/MCprep/blob/f774e80150621b1977703f05367a2d6ab96c6ed0/MCprep_addon/util.py#L116 and the console prints `Running vertex group: influence` multiple times.

# Why am I doing this?
I want to take a world from Minecraft, export it into a 3D engine format and then be able to import it into Unity for creating cinematic's and learning about Unity's new features regarding that. The world is quite massive but I'm already planning some optimizations for when it reaches Unity

# Versions
Blender 2.8
Windows 10
MCPrep 3.2.1

# Steps to reproduce
Create a new Blender file and open the scripting tab, paste the following (very poorly written) Python code (I've never used Python before so this is literally my first ever Python script, it's also not meant to be proper code as it's supposed to be just a quick botch to get to the next stage for me)

You would need the source obj/mtl files though to be able to re-produce this however I don't see why this wouldn't be able to be reproduced on a single obj file (just ensure that it has a lot of tall grass, that seems to cause the issue)

    import bpy
    from bpy import context
    from time import sleep
    import threading
    import gc

    startX = 1584*16 # original start was 24672, working in manual batches atm
    startZ = -192
    endX = 25888
    endZ = 768

    offsetX = -startX;
    offsetZ = -startZ;

    chunkStartX = int(startX / 16);
    chunkStartZ = int(startZ / 16);
    chunkEndX = int(endX / 16);
    chunkEndZ = int(endZ / 16);

    chunkDeltaX = int(abs(chunkStartX - chunkEndX));
    chunkDeltaZ = int(abs(chunkStartZ - chunkEndZ));

    totalChunks = chunkDeltaX * chunkDeltaZ;

    chunkX = chunkStartX
    chunkZ = chunkStartZ

    def clearAll():
        bpy.ops.object.select_all(action='SELECT')
        
        for o in bpy.context.selected_objects:
            for m in o.modifiers:
                o.modifiers.remove(m)
        
        bpy.ops.object.delete(use_global=True)
        bpy.ops.object.select_all(action='DESELECT')
        
        context = bpy.context
        scene = context.scene

        for c in scene.collection.children:
            scene.collection.children.unlink(c)
        
        for c in bpy.data.collections:
            bpy.data.collections.remove(c)
        
        for material in bpy.data.materials:
            material.user_clear()
            bpy.data.materials.remove(material)
            
        for texture in bpy.data.textures:
            texture.user_clear()
            bpy.data.textures.remove(texture)
            
        for block in bpy.data.meshes:
            block.user_clear()
            bpy.data.meshes.remove(block)
        
        for block in bpy.data.images:
            block.user_clear()
            bpy.data.images.remove(block)
        
        for block in bpy.data.objects:
            block.user_clear()
            bpy.data.objects.remove(block)
        
        gc.collect()
        bpy.ops.script.reload()

    clearAll()

    def incrementCounter():
        global chunkZ, chunkX
        chunkZ += 1
        if chunkZ >= chunkEndZ:
            chunkZ = chunkStartZ
            chunkX += 1
                
    def processChunks():
        file_loc = 'C:\\Users\\MegaMiley\\Desktop\\Exporter\\chunk_{}_{}.obj'.format(chunkX, chunkZ)
        print(file_loc);
        imported_object = bpy.ops.import_scene.obj(filepath=file_loc)

        bpy.ops.object.select_all(action='SELECT')
        
        try:
            bpy.ops.mcprep.prep_materials(animateTextures=False, autoFindMissingTextures=True, combineMaterials=True, improveUiSettings=True, usePrincipledShader=True, useReflections=True, useExtraMaps=True, makeSolid=False)
        except Exception:
            pass
        
        try:
            bpy.ops.mcprep.meshswap(meshswap_join=False, prep_materials=False)
        except Exception:
            pass

        bpy.ops.export_scene.fbx(bake_anim=False, filepath="C:\\Users\\MegaMiley\\Desktop\\Exporter\\fbx\\chunk_{}_{}.fbx".format(chunkX, chunkZ))
        
        clearAll()

    for i in range(chunkDeltaZ):
        processChunks()
        incrementCounter()

The `clearAll` function is there to clear out any data that may linger after processing a chunk (without this it would quickly shoot towards using up 32Gb of RAM). My assumption at this point is that there is some kind of leak in the `vertex_groups` where they don't get cleared out (same link as above). I'm hoping it's something in my script but at this point I am unable to improve the memory footprint of my script any further (RAM usage stays quite constant at 230Mb)

# Extra info
As soon as the script stops and starts again this issue is gone, it only exists whilst this script is running and gets worse the longer it's running, this is also the reason for the odd `incrementCounter` function and the for loop at the end of the script.

# Resolution
Did I make a silly mistake here or is there potentially a leak in the MCPrep code? And what could be done to resolve the lag issue?

**Comment by TheDuckCow** [comment]
Hi there! Firstly, thank you for clearly and cleanly stating your intents, it helps a lot. Secondly, a gut reaction I have is - how helpful is MCprep as an intermediate for you? I suppose Blender itself is helpful for generating the FBXs, and then I perhaps can understand some beneficial material fixes that get exported. 

But meshswap, well I have to say, I'm honestly surprised you've run into some issues. My equivalent of your first python code here is in fact the meshswap code. Old, clunky, and makes me twitch just looking at it because of all the case conditionals. I can't help but offer a suggestion to see if you can do the meshswapping work upfront, especially if it is for only very specific blocks. Jmc2obj does provide some options for specifying obj files to already export the objects desired up front, doing it downstream in something like MCprep will always have some artifacts and be substantially slower. 

## However, I'll still see if I can help out with the issue and problem as stated. Here are my initial thoughts:

- Gut reaction: memory issues are not surprising. Python does not offer very rigid memory management (ie releasing old memory immediately even after a function call), so there may not be a ton we can do there besides avoiding using such amount of memory in the first place
- But, I do believe there are many places that likely the meshswap code could be made more efficient, memory and cpu wise alike; however, I've avoided touching it unless the impact of a change is straightforward and obvious to not have any other side effects (again, that bit of code is a mess).
- As for the specific reference to `def obj_copy`, in order to deep copy the objects, blender has to create complete memory copies of the data in question, and for each object. Blender generally tends to not fare well with high number of objects, and because MeshSwap iterates on a block by block basis, this would definitely add up over time
- Even if the objects are later merged together, references hang around for a little bit to support undo/redo. This may be a notable factor here, as the "undo" operation has to essentially keep track of all the sequence of actions or at least the prior state to revert back to.
- An optimization I've long been thinking about, which would severely improve this situation, is actually using dupliverts as an intermediate. More on this below.
- Some thoughts on your specific code:
  - bpy.ops.object.select_all(action='SELECT'), is this right? Are you effectively re-meshswapping (or, at least checking) every chunk? if so, this could explain a lot. See if you can proactively only select the newly imported objects, I beleive that's actually the default behavior so just make a point to *deselect* things ahead of time
  - On the note of deselection, avoid operators where you can - they actually run slower than direct python calls. See [this resource page](https://theduckcow.com/2019/update-addons-both-blender-28-and-27-support/), but in the case here, just use obj.get_select(False) in a loop over all objects
  - This applies even if you are deleting things, which I see now you are attempting to do between each chunk process; again also see if lowering the 
  - As a quicker as-is workaround, what about running this from the command line? You could essentially run a command like so: `blender -p my_script.py` and then within this script, skip all the "clear all stuff" and just literally run the new file operator, and truly start fresh between each chunk. I'm not confident that clearing all as you have written is really going to work as you hope, even the garbage collection is a soft contract from what I understand.


## Back to the point on a potential future optimization:

Current state process
- Go through selected objects (could be a whole chunk as one object), split out by material type (ie ≈ minecraft block type)
- Use manual logic to find all locations of a block instance in the raw export; save an overall list of these center points
- Import the meshswap object
- Duplicate this meshswap object to each vertex location (+ do other things like adjust rotation, location variance for grass etc)
- Delete the original block vertices from the raw export
- Join the generated meshswap blocks together (if option selected).

Possible optimization
- Go through selected objects
- Use manual logic to find all locations of a block instance...
- Use manual logic to find all locations of a block instance in the raw export; save an overall list of these center points
- Import the meshswap object
- Generate a single "mesh" of vertices for each location to replicate the object, and use *dupliverts* to 
- "Make real" on dupliverts to convert these into a single mesh

I think this will make things quite a bit faster, and save on memory as it leans more on blenders C-level instancing rather than memory cost of python performing the deep copies of objects. However, I have to account for additional cases such as rotations (which cannot be simply accounted for in dupligroups). All in all, I would not proceed depending on this being ready in the near future, but it has been on my mind to do.

Hope this helps!

**Comment by MileyHollenberg** [comment]
Thank you for your reply, there’s a ton of useful information in here! 

Yes, MCPrep is a great tool as an intermediary for my use-case, I was originally planning on writing a custom Minecraft Importer Editor Extension for Unity but jMC2obj combined with MCPrep would basically do the same thing without me having to write all that code.

As for my code, more than half of the Blender specific calls were simply copy pasted from the scripting console when executing those actions manually. I’ve not tried running the code through the command line because I simply didn’t know of that possibility yet plus my assumption would be that the time to reload Blender through the command line would have a negative impact on the overal time as well (though if I split it up into chunk rows this might not be too much of an issue)

For the other parts I will take a look at them later tonight when I get back home from work

Also a quick bit of guidance on how to potentially improve the ‘meshswap’ logic.
This is assuming that you have to process each face within the model fyi, in Minecraft each chunk isn’t processed as its 16x16x256 block of data but instead it’s split up into 16x16x16 sub chunks, this allows the game to perform less optimal checks on a large scale without a significant slowdown (I’ve forgotten the specific name of this type of optimization), employing such a system to your ‘meshswap‘ code could greatly improve the performance without requiring a full re-imagination of how to even check for these specific meshes

Again thank you for your reply and also thank you for making this tool available to everyone!

**Comment by MileyHollenberg** [comment]
Quick update, the script finally finished and I'm now importing it into Unity. I will be focusing on not making Unity grind to a halt first before re-visiting this script as I'm most likely going to need it in the future (though on a smaller scale hopefully)

**Comment by TheDuckCow** [comment]
Glad to hear it's progress.

Thanks for the suggestion on optimizing meshswap. At the moment, indeed I am running checks face-by-face, but I already know that every single face is relevant (because a previous step separated out materials into separate objects). And so every face is useful to add into the matrix of interpreted block locations. Thus, unfortunately, I don't have a way really to 'skip' what faces we look at. I _do_ potentially have the option to try and run this process in parallel threads perhaps, especially if I am able to do that within the same material object. That plus other general optimizations such as what I mentioned about spawning via instances instead of full object copies could all help to make it significantly more optimal.

**Comment by TheDuckCow** [comment]
Curious, is this a project you are still working on? I'm curious to hear if any of the changes I made to meshswapping recently will make a difference here or not. Several things were optimized, though the "vertex copy" piece itself I don't think I really touched.

**Comment by MileyHollenberg** [comment]
I basically abandoned the project.. (I do that way too often with my personal projects) so no progress was made on this 😅

**Comment by TheDuckCow** [comment]
Fair enough! Closing this in the meantime but can always reopen if you come back to this.
